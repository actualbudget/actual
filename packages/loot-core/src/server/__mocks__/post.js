export {
  handleRequest as post,
  handleRequestBinary as postBinary,
} from '../tests/mockSyncServer';

export const get = function (url) {
  throw new Error('get unimplemented');
};
