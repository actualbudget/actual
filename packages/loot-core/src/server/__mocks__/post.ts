// @ts-strict-ignore
export {
  handleRequest as post,
  handleRequestBinary as postBinary,
} from '../tests/mockSyncServer';

export const get = function () {
  throw new Error('get unimplemented');
};
