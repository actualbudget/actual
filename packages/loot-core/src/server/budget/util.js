import { number } from '../spreadsheet/globals';

export { number } from '../spreadsheet/globals';

export function sumAmounts(...amounts) {
  return amounts.reduce((total, amount) => {
    return total + number(amount);
  }, 0);
}

export function flatten2(arr) {
  return Array.prototype.concat.apply([], arr);
}

export function unflatten2(arr) {
  let res = [];
  for (let i = 0; i < arr.length; i += 2) {
    res.push([arr[i], arr[i + 1]]);
  }
  return res;
}

// Note that we don't restrict values to `Number.MIN_SAFE_INTEGER <= value <= Number.MAX_SAFE_INTEGER`
// where `Number.MAX_SAFE_INTEGER == 2^53 - 1` but a smaller range over `-(2^43-1) <= value <= 2^43 - 1`.
// This ensure that the number is accurate not just for the integer component but for 3 decimal places also.
//
// This gives us the guarantee that can use `safeNumber` on number whether they are unscaled user inputs
// or they have been converted to integers (using `amountToInteger`).

const MAX_SAFE_NUMBER = 2 ** 43 - 1;
const MIN_SAFE_NUMBER = -MAX_SAFE_NUMBER;

export function safeNumber(value) {
  value = number(value);
  if (value > MAX_SAFE_NUMBER || value < MIN_SAFE_NUMBER) {
    throw new Error(
      "Can't safely perform arithmetic with number: " + JSON.stringify(value)
    );
  }
  return value;
}
