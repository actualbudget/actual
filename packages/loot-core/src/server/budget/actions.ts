import * as monthUtils from '../../shared/months';
import { safeNumber } from '../../shared/util';
import * as db from '../db';
import * as prefs from '../prefs';
import * as sheet from '../sheet';
import { batchMessages } from '../sync';

export async function getSheetValue(
  sheetName: string,
  cell: string,
): Promise<number> {
  const node = await sheet.getCell(sheetName, cell);
  return safeNumber(typeof node.value === 'number' ? node.value : 0);
}

// We want to only allow the positive movement of money back and
// forth. buffered should never be allowed to go into the negative,
// and you shouldn't be allowed to pull non-existant money from
// leftover.
function calcBufferedAmount(
  toBudget: number,
  buffered: number,
  amount: number,
): number {
  amount = Math.min(Math.max(amount, -buffered), Math.max(toBudget, 0));
  return buffered + amount;
}

function getBudgetTable(): string {
  let { budgetType } = prefs.getPrefs() || {};
  return budgetType === 'report' ? 'reflect_budgets' : 'zero_budgets';
}

export function isReflectBudget(): boolean {
  let { budgetType } = prefs.getPrefs();
  return budgetType === 'report';
}

function dbMonth(month: string): number {
  return parseInt(month.replace('-', ''));
}

// TODO: complete list of fields.
type BudgetData = {
  is_income: 1 | 0;
  category: string;
  amount: number;
};

function getBudgetData(table: string, month: string): Promise<BudgetData[]> {
  return db.all(
    `
    SELECT b.*, c.is_income FROM v_categories c
    LEFT JOIN ${table} b ON b.category = c.id
    WHERE c.tombstone = 0 AND b.month = ?
  `,
    [month],
  );
}

function getAllMonths(startMonth: string): string[] {
  let { createdMonths } = sheet.get().meta();
  let latest = null;
  for (let month of createdMonths) {
    if (latest == null || month > latest) {
      latest = month;
    }
  }
  return monthUtils.rangeInclusive(startMonth, latest);
}

// TODO: Valid month format in all the functions below

export function getBudget({
  category,
  month,
}: {
  category: string;
  month: string;
}): number {
  let table = getBudgetTable();
  let existing = db.firstSync(
    `SELECT * FROM ${table} WHERE month = ? AND category = ?`,
    [dbMonth(month), category],
  );
  return existing ? existing.amount || 0 : 0;
}

export function setBudget({
  category,
  month,
  amount,
}: {
  category: string;
  month: string;
  amount: unknown;
}): Promise<void> {
  amount = safeNumber(typeof amount === 'number' ? amount : 0);
  const table = getBudgetTable();

  let existing = db.firstSync(
    `SELECT id FROM ${table} WHERE month = ? AND category = ?`,
    [dbMonth(month), category],
  );
  if (existing) {
    return db.update(table, { id: existing.id, amount });
  }
  return db.insert(table, {
    id: `${dbMonth(month)}-${category}`,
    month: dbMonth(month),
    category,
    amount,
  });
}

export function setBuffer(month: string, amount: unknown): Promise<void> {
  let existing = db.firstSync(
    `SELECT id FROM zero_budget_months WHERE id = ?`,
    [month],
  );
  if (existing) {
    return db.update('zero_budget_months', {
      id: existing.id,
      buffered: amount,
    });
  }
  return db.insert('zero_budget_months', { id: month, buffered: amount });
}

function setCarryover(
  table: string,
  category: string,
  month: string,
  flag: boolean,
): Promise<void> {
  let existing = db.firstSync(
    `SELECT id FROM ${table} WHERE month = ? AND category = ?`,
    [month, category],
  );
  if (existing) {
    return db.update(table, { id: existing.id, carryover: flag ? 1 : 0 });
  }
  return db.insert(table, {
    id: `${month}-${category}`,
    month,
    category,
    carryover: flag ? 1 : 0,
  });
}

// Actions

export async function copyPreviousMonth({
  month,
}: {
  month: string;
}): Promise<void> {
  let prevMonth = dbMonth(monthUtils.prevMonth(month));
  let table = getBudgetTable();
  let budgetData = await getBudgetData(table, prevMonth.toString());

  await batchMessages(async () => {
    budgetData.forEach(prevBudget => {
      if (prevBudget.is_income === 1 && !isReflectBudget()) {
        return;
      }
      setBudget({
        category: prevBudget.category,
        month,
        amount: prevBudget.amount,
      });
    });
  });
}

export async function copySinglePreviousMonth({
  month,
  category,
}: {
  month: string;
  category: string;
}): Promise<void> {
  let prevMonth = monthUtils.prevMonth(month);
  let newAmount = await getSheetValue(
    monthUtils.sheetForMonth(prevMonth),
    'budget-' + category,
  );
  await batchMessages(async () => {
    setBudget({ category: category, month, amount: newAmount });
  });
}

export async function setZero({ month }: { month: string }): Promise<void> {
  let categories = await db.all(
    'SELECT * FROM v_categories WHERE tombstone = 0',
  );

  await batchMessages(async () => {
    categories.forEach(cat => {
      if (cat.is_income === 1 && !isReflectBudget()) {
        return;
      }
      setBudget({ category: cat.id, month, amount: 0 });
    });
  });
}

export async function set3MonthAvg({
  month,
}: {
  month: string;
}): Promise<void> {
  let categories = await db.all(
    'SELECT * FROM v_categories WHERE tombstone = 0',
  );

  let prevMonth1 = monthUtils.prevMonth(month);
  let prevMonth2 = monthUtils.prevMonth(prevMonth1);
  let prevMonth3 = monthUtils.prevMonth(prevMonth2);

  await batchMessages(async () => {
    for (let cat of categories) {
      if (cat.is_income === 1 && !isReflectBudget()) {
        continue;
      }

      let spent1 = await getSheetValue(
        monthUtils.sheetForMonth(prevMonth1),
        'sum-amount-' + cat.id,
      );
      let spent2 = await getSheetValue(
        monthUtils.sheetForMonth(prevMonth2),
        'sum-amount-' + cat.id,
      );
      let spent3 = await getSheetValue(
        monthUtils.sheetForMonth(prevMonth3),
        'sum-amount-' + cat.id,
      );

      const avg = Math.round((spent1 + spent2 + spent3) / 3);
      setBudget({ category: cat.id, month, amount: -avg });
    }
  });
}

export async function setNMonthAvg({
  month,
  N,
  category,
}: {
  month: string;
  N: number;
  category: string;
}): Promise<void> {
  let prevMonth = monthUtils.prevMonth(month);
  let sumAmount = 0;
  for (let l = 0; l < N; l++) {
    sumAmount += await getSheetValue(
      monthUtils.sheetForMonth(prevMonth),
      'sum-amount-' + category,
    );
    prevMonth = monthUtils.prevMonth(prevMonth);
  }
  await batchMessages(async () => {
    const avg = Math.round(sumAmount / N);
    setBudget({ category: category, month, amount: -avg });
  });
}

export async function holdForNextMonth({
  month,
  amount,
}: {
  month: string;
  amount: number;
}): Promise<boolean> {
  let row = await db.first(
    'SELECT buffered FROM zero_budget_months WHERE id = ?',
    [month],
  );

  let sheetName = monthUtils.sheetForMonth(month);
  let toBudget = await getSheetValue(sheetName, 'to-budget');

  if (toBudget > 0) {
    let bufferedAmount = calcBufferedAmount(
      toBudget,
      (row && row.buffered) || 0,
      amount,
    );

    await setBuffer(month, bufferedAmount);
    return true;
  }
  return false;
}

export async function resetHold({ month }: { month: string }): Promise<void> {
  await setBuffer(month, 0);
}

export async function coverOverspending({
  month,
  to,
  from,
}: {
  month: string;
  to: string;
  from: string;
}): Promise<void> {
  let sheetName = monthUtils.sheetForMonth(month);
  let toBudgeted = await getSheetValue(sheetName, 'budget-' + to);
  let leftover = await getSheetValue(sheetName, 'leftover-' + to);
  let leftoverFrom = await getSheetValue(
    sheetName,
    from === 'to-be-budgeted' ? 'to-budget' : 'leftover-' + from,
  );

  if (leftover >= 0 || leftoverFrom <= 0) {
    return;
  }

  let amountCovered = Math.min(-leftover, leftoverFrom);

  // If we are covering it from the to be budgeted amount, ignore this
  if (from !== 'to-be-budgeted') {
    const fromBudgeted = await getSheetValue(sheetName, 'budget-' + from);
    await setBudget({
      category: from,
      month,
      amount: fromBudgeted - amountCovered,
    });
  }

  await setBudget({ category: to, month, amount: toBudgeted + amountCovered });
}

export async function transferAvailable({
  month,
  amount,
  category,
}: {
  month: string;
  amount: number;
  category: string;
}): Promise<void> {
  let sheetName = monthUtils.sheetForMonth(month);
  let leftover = await getSheetValue(sheetName, 'to-budget');
  amount = Math.max(Math.min(amount, leftover), 0);

  let budgeted = await getSheetValue(sheetName, 'budget-' + category);
  await setBudget({ category, month, amount: budgeted + amount });
}

export async function transferCategory({
  month,
  amount,
  from,
  to,
}: {
  month: string;
  amount: number;
  to: string;
  from: string;
}): Promise<void> {
  const sheetName = monthUtils.sheetForMonth(month);
  const fromBudgeted = await getSheetValue(sheetName, 'budget-' + from);

  await setBudget({ category: from, month, amount: fromBudgeted - amount });

  // If we are simply moving it back into available cash to budget,
  // don't do anything else
  if (to !== 'to-be-budgeted') {
    const toBudgeted = await getSheetValue(sheetName, 'budget-' + to);
    await setBudget({ category: to, month, amount: toBudgeted + amount });
  }
}

export async function setCategoryCarryover({
  startMonth,
  category,
  flag,
}: {
  startMonth: string;
  category: string;
  flag: boolean;
}): Promise<void> {
  let table = getBudgetTable();
  let months = getAllMonths(startMonth);

  await batchMessages(async () => {
    for (let month of months) {
      setCarryover(table, category, dbMonth(month).toString(), flag);
    }
  });
}
