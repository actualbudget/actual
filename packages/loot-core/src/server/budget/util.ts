import { safeNumber } from '../../shared/util';
import { number } from '../spreadsheet/globals';

export { number } from '../spreadsheet/globals';

export function sumAmounts(...amounts) {
  return safeNumber(
    amounts.reduce((total, amount) => {
      return total + number(amount);
    }, 0),
  );
}

export function flatten2(arr) {
  return Array.prototype.concat.apply([], arr);
}

export function unflatten2(arr) {
  let res = [];
  for (let i = 0; i < arr.length; i += 2) {
    res.push([arr[i], arr[i + 1]]);
  }
  return res;
}
