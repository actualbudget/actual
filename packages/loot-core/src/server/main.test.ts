import { getClock, deserializeClock } from '@actual-app/crdt';

import { expectSnapshotWithDiffer } from '../mocks/util';
import * as connection from '../platform/server/connection';
import * as fs from '../platform/server/fs';
import * as monthUtils from '../shared/months';

import * as budgetActions from './budget/actions';
import * as budget from './budget/base';
import * as db from './db';
import { handlers } from './main';
import {
  runHandler,
  runMutator,
  disableGlobalMutations,
  enableGlobalMutations,
} from './mutators';
import { post } from './post';
import * as prefs from './prefs';
import * as sheet from './sheet';

jest.mock('./post');

beforeEach(async () => {
  await global.emptyDatabase()();
  disableGlobalMutations();
});

afterEach(async () => {
  await runHandler(handlers['close-budget']);
  connection.resetEvents();
  enableGlobalMutations();
  global.currentMonth = null;
});

async function createTestBudget(name) {
  let templatePath = fs.join(__dirname, '/../mocks/files', name);
  let budgetPath = fs.join(__dirname, '/../mocks/files/budgets/test-budget');
  fs._setDocumentDir(fs.join(budgetPath, '..'));

  await fs.mkdir(budgetPath);
  await fs.copyFile(
    fs.join(templatePath, 'metadata.json'),
    fs.join(budgetPath, 'metadata.json'),
  );
  await fs.copyFile(
    fs.join(templatePath, 'db.sqlite'),
    fs.join(budgetPath, 'db.sqlite'),
  );
}

describe('Budgets', () => {
  afterEach(async () => {
    fs._setDocumentDir(null);
    let budgetPath = fs.join(__dirname, '/../mocks/files/budgets/test-budget');

    if (await fs.exists(budgetPath)) {
      await fs.removeDirRecursively(budgetPath);
    }
  });

  test('budget is successfully loaded', async () => {
    await createTestBudget('default-budget-template');

    // Grab the clock to compare later
    await db.openDatabase('test-budget');
    let row = await db.first('SELECT * FROM messages_clock');

    let { error } = await runHandler(handlers['load-budget'], {
      id: 'test-budget',
    });
    expect(error).toBe(undefined);

    // Make sure the prefs were loaded
    expect(prefs.getPrefs().id).toBe('test-budget');

    // Make sure the clock has been loaded
    expect(getClock()).toEqual(deserializeClock(row.clock));
  });

  test('budget detects out of sync migrations', async () => {
    await createTestBudget('default-budget-template');

    await db.openDatabase('test-budget');
    await db.runQuery('INSERT INTO __migrations__ (id) VALUES (1000)');

    const spy = jest.spyOn(console, 'warn').mockImplementation();

    let { error } = await runHandler(handlers['load-budget'], {
      id: 'test-budget',
    });
    // There should be an error and the budget should be unloaded
    expect(error).toBe('out-of-sync-migrations');
    expect(db.getDatabase()).toBe(null);
    expect(prefs.getPrefs()).toBe(null);

    spy.mockRestore();
  });
});

describe('Accounts', () => {
  test('create accounts with correct starting balance', async () => {
    prefs.loadPrefs();
    prefs.savePrefs({ groupId: 'group' });

    await runMutator(async () => {
      // An income category is required because the starting balance is
      // categorized to it. Create one now.
      await db.insertCategoryGroup({
        id: 'group1',
        name: 'income',
        is_income: 1,
      });
      await db.insertCategory({
        name: 'income',
        cat_group: 'group1',
        is_income: 1,
      });
    });

    // Get accounts from the server. This isn't the normal API call,
    // we know that the mock server just returns hardcoded accounts
    let { accounts } = await post('/plaid/accounts', {});

    // Create the accounts for the bank (bank is generally ignored in tests)
    await runHandler(handlers['accounts-connect'], {
      institution: { institution_id: 1, name: 'Jamesy Bank' },
      publicToken: 'foo',
      accountIds: accounts.map(acct => acct.account_id),
    });

    // Import transactions for all accounts
    await runHandler(handlers['accounts-sync'], {});

    // Go through each account and make sure the starting balance was
    // created correctly
    const res = await db.all('SELECT * FROM accounts');
    for (let account of res) {
      const sum = await db.first(
        'SELECT sum(amount) as sum FROM transactions WHERE acct = ? AND starting_balance_flag = 0',
        [account.id],
      );
      const starting = await db.first(
        'SELECT * FROM transactions WHERE acct = ? AND starting_balance_flag = 1',
        [account.id],
      );
      expect(account.balance_current - sum.sum).toBe(starting.amount);

      // Also ensure that the starting balance has the earliest date
      // possible
      const earliestTrans = await db.first(
        'SELECT p.name as payee_name FROM transactions t LEFT JOIN payees p ON p.id = t.description WHERE acct = ? ORDER BY date LIMIT 1',
        [account.id],
      );
      expect(earliestTrans.payee_name).toBe('Starting Balance');
    }
  });

  test('Transfers are properly updated', async () => {
    await runMutator(async () => {
      await db.insertAccount({ id: 'one', name: 'one' });
      await db.insertAccount({ id: 'two', name: 'two' });
      await db.insertAccount({ id: 'three', name: 'three' });
      await db.insertPayee({
        id: 'transfer-one',
        name: '',
        transfer_acct: 'one',
      });
      await db.insertPayee({
        id: 'transfer-two',
        name: '',
        transfer_acct: 'two',
      });
      await db.insertPayee({
        id: 'transfer-three',
        name: '',
        transfer_acct: 'three',
      });
    });

    const id = 'test-transfer';
    await runHandler(handlers['transaction-add'], {
      id,
      account: 'one',
      amount: 5000,
      payee: 'transfer-two',
      date: '2017-01-01',
    });
    let differ = expectSnapshotWithDiffer(
      await db.all('SELECT * FROM transactions'),
    );

    let transaction = await db.getTransaction(id);
    await runHandler(handlers['transaction-update'], {
      ...(await db.getTransaction(id)),
      payee: 'transfer-three',
      date: '2017-01-03',
    });
    differ.expectToMatchDiff(await db.all('SELECT * FROM transactions'));

    transaction = await db.getTransaction(id);
    await runHandler(handlers['transaction-delete'], transaction);
    differ.expectToMatchDiff(await db.all('SELECT * FROM transactions'));
  });
});

describe('Budget', () => {
  test('new budgets should be created', async () => {
    const spreadsheet = await sheet.loadSpreadsheet(db);

    await runMutator(async () => {
      await db.insertCategoryGroup({
        id: 'incomeGroup',
        name: 'incomeGroup',
        is_income: 1,
      });
      await db.insertCategoryGroup({ id: 'group1', name: 'group1' });
      await db.insertCategory({ name: 'foo', cat_group: 'group1' });
      await db.insertCategory({ name: 'bar', cat_group: 'group1' });
    });

    let bounds = await runHandler(handlers['get-budget-bounds']);
    expect(bounds.start).toBe('2016-10');
    expect(bounds.end).toBe('2018-01');
    expect(spreadsheet.meta().createdMonths).toMatchSnapshot();

    // Add a transaction (which needs an account) earlier then the
    // current earliest budget to test if it creates the necessary
    // budgets for the earlier months
    await db.runQuery("INSERT INTO accounts (id, name) VALUES ('one', 'boa')");
    await runHandler(handlers['transaction-add'], {
      date: '2016-05-06',
      amount: 50,
      account: 'one',
    });

    // Fast-forward in time to a future month and make sure it creates
    // budgets for the months in the future
    global.currentMonth = '2017-02';

    bounds = await runHandler(handlers['get-budget-bounds']);
    expect(bounds.start).toBe('2016-02');
    expect(bounds.end).toBe('2018-02');
    expect(spreadsheet.meta().createdMonths).toMatchSnapshot();

    await new Promise(resolve => spreadsheet.onFinish(resolve));
  });

  test('budget updates when changing a category', async () => {
    const spreadsheet = await sheet.loadSpreadsheet(db);
    function captureChangedCells(func) {
      return new Promise<unknown[]>(async resolve => {
        let changed = [];
        let remove = spreadsheet.addEventListener('change', ({ names }) => {
          changed = changed.concat(names);
        });
        await func();
        remove();
        spreadsheet.onFinish(() => {
          resolve(changed);
        });
      });
    }

    // Force the system to start tracking these months so budgets are
    // automatically updated when adding/deleting categories
    await db.runQuery('INSERT INTO created_budgets (month) VALUES (?)', [
      '2017-01',
    ]);
    await db.runQuery('INSERT INTO created_budgets (month) VALUES (?)', [
      '2017-02',
    ]);
    await db.runQuery('INSERT INTO created_budgets (month) VALUES (?)', [
      '2017-03',
    ]);
    await db.runQuery('INSERT INTO created_budgets (month) VALUES (?)', [
      '2017-04',
    ]);

    let categories;
    await captureChangedCells(async () => {
      await runMutator(() =>
        db.insertCategoryGroup({ id: 'group1', name: 'group1' }),
      );
      categories = [
        await runHandler(handlers['category-create'], {
          name: 'foo',
          groupId: 'group1',
        }),
        await runHandler(handlers['category-create'], {
          name: 'bar',
          groupId: 'group1',
        }),
        await runHandler(handlers['category-create'], {
          name: 'baz',
          groupId: 'group1',
        }),
        await runHandler(handlers['category-create'], {
          name: 'biz',
          groupId: 'group1',
        }),
      ];
    });

    await db.runQuery("INSERT INTO accounts (id, name) VALUES ('boa', 'boa')");
    let trans = {
      id: 'boa-transaction',
      date: '2017-02-06',
      amount: 5000,
      account: 'boa',
      category: categories[0],
    };
    // Test insertions
    let changed = await captureChangedCells(() =>
      runHandler(handlers['transaction-add'], trans),
    );
    expect(changed.sort()).toMatchSnapshot();
    // Test updates
    changed = await captureChangedCells(async () => {
      await runHandler(handlers['transaction-update'], {
        ...(await db.getTransaction(trans.id)),
        amount: 7000,
      });
    });
    expect(changed.sort()).toMatchSnapshot();
    // Test deletions
    changed = await captureChangedCells(async () => {
      await runHandler(handlers['transaction-delete'], { id: trans.id });
    });
    expect(changed.sort()).toMatchSnapshot();
  });
});

describe('Categories', () => {
  test('can be deleted', async () => {
    await sheet.loadSpreadsheet(db);

    await runMutator(async () => {
      await db.insertCategoryGroup({ id: 'group1', name: 'group1' });
      await db.insertCategory({ id: 'foo', name: 'foo', cat_group: 'group1' });
      await db.insertCategory({ id: 'bar', name: 'bar', cat_group: 'group1' });
    });

    let categories = await db.getCategories();
    expect(categories.length).toBe(2);
    expect(categories.find(cat => cat.name === 'foo')).not.toBeNull();
    expect(categories.find(cat => cat.name === 'bar')).not.toBeNull();
    await runHandler(handlers['category-delete'], { id: 'foo' });

    categories = await db.getCategories();
    expect(categories.length).toBe(1);
    expect(categories.find(cat => cat.name === 'bar')).not.toBeNull();
  });

  test('transfers properly when deleted', async () => {
    await sheet.loadSpreadsheet(db);

    let transId = await runMutator(async () => {
      await db.insertCategoryGroup({ id: 'group1', name: 'group1' });
      await db.insertCategoryGroup({ id: 'group1b', name: 'group1b' });
      await db.insertCategoryGroup({
        id: 'group2',
        name: 'group2',
        is_income: 1,
      });
      await db.insertCategory({ id: 'foo', name: 'foo', cat_group: 'group1' });
      await db.insertCategory({ id: 'bar', name: 'bar', cat_group: 'group1b' });
      await db.insertCategory({
        id: 'income1',
        name: 'income1',
        is_income: 1,
        cat_group: 'group2',
      });
      await db.insertCategory({
        id: 'income2',
        name: 'income2',
        is_income: 1,
        cat_group: 'group2',
      });

      return await db.insertTransaction({
        date: '2017-01-01',
        account: 'acct',
        amount: 4500,
        category: 'foo',
      });
    });

    await budget.createAllBudgets();

    // Set a budget value for the category `foo` of 1000
    let sheetName = monthUtils.sheetForMonth('2018-01');
    await budgetActions.setBudget({
      category: 'foo',
      month: '2018-01',
      amount: 1000,
    });
    expect(sheet.getCellValue(sheetName, 'group-budget-group1')).toBe(1000);
    expect(sheet.getCellValue(sheetName, 'group-budget-group1b')).toBe(0);

    // Make sure the transaction has a category of `foo`
    let trans = await db.getTransaction(transId);
    expect(trans.category).toBe('foo');

    await runHandler(handlers['category-delete'], {
      id: 'foo',
      transferId: 'bar',
    });

    // Make sure the transaction has been updated
    trans = await db.getTransaction(transId);
    expect(trans.category).toBe('bar');

    // Make sure the budget value was transferred
    expect(sheet.getCellValue(sheetName, 'group-budget-group1')).toBe(0);
    expect(sheet.getCellValue(sheetName, 'group-budget-group1b')).toBe(1000);

    // Transfering an income category to an expense just doesn't make
    // sense. Make sure this doesn't do anything.
    let { error } = await runHandler(handlers['category-delete'], {
      id: 'income1',
      transferId: 'bar',
    });
    expect(error).toBe('category-type');

    let categories = await db.getCategories();
    expect(categories.find(cat => cat.id === 'income1')).toBeDefined();

    // Make sure you can delete income categories
    await runHandler(handlers['category-delete'], {
      id: 'income1',
      transferId: 'income2',
    });

    categories = await db.getCategories();
    expect(categories.find(cat => cat.id === 'income1')).not.toBeDefined();
  });
});
