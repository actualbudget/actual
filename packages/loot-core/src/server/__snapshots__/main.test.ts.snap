// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accounts Transfers are properly updated 1`] = `
Array [
  Object {
    "acct": "one",
    "amount": 5000,
    "category": null,
    "cleared": 1,
    "date": 20170101,
    "description": "transfer-two",
    "error": null,
    "financial_id": null,
    "id": "test-transfer",
    "imported_description": null,
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": "id2",
    "type": null,
  },
  Object {
    "acct": "two",
    "amount": -5000,
    "category": null,
    "cleared": 0,
    "date": 20170101,
    "description": "transfer-one",
    "error": null,
    "financial_id": null,
    "id": "id2",
    "imported_description": null,
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": "test-transfer",
    "type": null,
  },
]
`;

exports[`Accounts Transfers are properly updated 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -2,12 +2,12 @@
    Object {
      \\"acct\\": \\"one\\",
      \\"amount\\": 5000,
      \\"category\\": null,
      \\"cleared\\": 1,
-     \\"date\\": 20170101,
-     \\"description\\": \\"transfer-two\\",
+     \\"date\\": 20170103,
+     \\"description\\": \\"transfer-three\\",
      \\"error\\": null,
      \\"financial_id\\": null,
      \\"id\\": \\"test-transfer\\",
      \\"imported_description\\": null,
      \\"isChild\\": 0,
@@ -22,15 +22,15 @@
      \\"tombstone\\": 0,
      \\"transferred_id\\": \\"id2\\",
      \\"type\\": null,
    },
    Object {
-     \\"acct\\": \\"two\\",
+     \\"acct\\": \\"three\\",
      \\"amount\\": -5000,
      \\"category\\": null,
      \\"cleared\\": 0,
-     \\"date\\": 20170101,
+     \\"date\\": 20170103,
      \\"description\\": \\"transfer-one\\",
      \\"error\\": null,
      \\"financial_id\\": null,
      \\"id\\": \\"id2\\",
      \\"imported_description\\": null,"
`;

exports[`Accounts Transfers are properly updated 3`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -17,12 +17,12 @@
      \\"parent_id\\": null,
      \\"pending\\": 0,
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
-     \\"tombstone\\": 0,
-     \\"transferred_id\\": \\"id2\\",
+     \\"tombstone\\": 1,
+     \\"transferred_id\\": null,
      \\"type\\": null,
    },
    Object {
      \\"acct\\": \\"three\\",
      \\"amount\\": -5000,
@@ -41,10 +41,10 @@
      \\"parent_id\\": null,
      \\"pending\\": 0,
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
-     \\"tombstone\\": 0,
+     \\"tombstone\\": 1,
      \\"transferred_id\\": \\"test-transfer\\",
      \\"type\\": null,
    },
  ]"
`;

exports[`Budget budget updates when changing a category 1`] = `
Array [
  "budget201702!sum-amount-id2",
]
`;

exports[`Budget budget updates when changing a category 2`] = `
Array [
  "budget201702!sum-amount-id2",
]
`;

exports[`Budget budget updates when changing a category 3`] = `
Array [
  "budget201702!sum-amount-id2",
]
`;

exports[`Budget new budgets should be created 1`] = `
Set {
  "2016-10",
  "2016-11",
  "2016-12",
  "2017-01",
  "2017-02",
  "2017-03",
  "2017-04",
  "2017-05",
  "2017-06",
  "2017-07",
  "2017-08",
  "2017-09",
  "2017-10",
  "2017-11",
  "2017-12",
  "2018-01",
}
`;

exports[`Budget new budgets should be created 2`] = `
Set {
  "2016-10",
  "2016-11",
  "2016-12",
  "2017-01",
  "2017-02",
  "2017-03",
  "2017-04",
  "2017-05",
  "2017-06",
  "2017-07",
  "2017-08",
  "2017-09",
  "2017-10",
  "2017-11",
  "2017-12",
  "2018-01",
  "2016-02",
  "2016-03",
  "2016-04",
  "2016-05",
  "2016-06",
  "2016-07",
  "2016-08",
  "2016-09",
  "2018-02",
}
`;
